vsintegration/src/FSharp.Editor/Common/CancellableTasks.fs

```diff
@@ -1119,6 +1119,14 @@ module CancellableTasks =
 
         let inline ignore ([<InlineIfLambda>] ctask: CancellableTask<_>) = toUnit ctask
 
+        let inline ifCanceledReturn value (ctask : CancellableTask<_>) =
+            cancellableTask {
+                try
+                    return! ctask
+                with :? OperationCanceledException ->
```

Comment:
We should only do that if the token causing the OCE is a different one from the one our cancellable uses.
Or, maybe a lot more explicit - out code failing on project options would throw any other exception, which will be explicitly caught here.

Reply:
> We should only do that if the token causing the OCE is a different one from the one our cancellable uses.

What token is used when we throw from our cancellable? Also how can we tell who caused it?

> Or, maybe a lot more explicit - out code failing on project options would throw any other exception, which will be explicitly caught here.

The downside of that is that we would have to start catching it in _a lot_ of places. 

---

Also, what actually happens when this is canceled for another reason? Since the OCE from here goes to Watson no one is relying on it in any way. Assuming the caller handles it somehow when they initiate the cancellation, they shouldn't have a problem with us returning an empty result (which will in that case probably just be thrown away).

Reply:
Ok, we could check that the `(CancellationToken we get called with).IsCancellationRequested = false` which seems to be the case when we just throw an OCE.

Reply:
OCE exception has a field with the token that caused it. Comparison of the two tokens should be possible.

Reply:
Btw. the fact that we are (sometimes) reporting OCE to Watson is a mistake caused by general catch handle on our end, we should not be doing that.

Reply:
> Btw. the fact that we are (sometimes) reporting OCE to Watson is a mistake caused by general catch handle on our end, we should not be doing that.

How is that? These get called through MEF. We're not catching anything. Or are we hooking into this mechanism somewhere indirectly?

Reply:
At some places, we have a universal try-with catching all exceptions, and doing our own telemetry reporting.

Reply:
Yeah we should deal with those. But it's not the case here.

Reply:
> OCE exception has a field with the token that caused it. Comparison of the two tokens should be possible.

I don't recommend this, Stephen Cleary just made a [blogpost on why](https://blog.stephencleary.com/2024/10/cancellation-6-linking.html).

Reply:
That sounds reasonable, thanks for the tip Jimmy @TheAngryByrd .
We can use .IsCancellationRequested check in the editor layer, since the most common APIs exported to Roslyn do have a cancellationToken parameter.